File: C++/soln-stack-problems/longest-valid-parentheses.cpp
/**
 @author Farheen Bano
  
 Reference-
 https://leetcode.com/problems/longest-valid-parentheses/
*/

class Solution {
public:
    int longestValidParentheses(string s) {
        stack<int> st;
        st.push(-1);
        int maxLen=0;
        for(int i=0;i<s.length();i++){
            if(s[i]=='('){
                st.push(i);
            }
            else{
                st.pop();
                if(st.empty()){
                    st.push(i);
                }
                else{
                    maxLen=max(maxLen,i-st.top());
                }
            }
        }
        return maxLen;
    }
};

File: Python/soln-stack-problems/longest-valid-parentheses.py
'''
 @author Farheen Bano
  
 Reference-
 https://leetcode.com/problems/longest-valid-parentheses/
'''

class Solution:
    def longestValidParentheses(self, s: str) -> int:
        stack=[]
        stack.append(-1)
        maxLen=0
        
        for i in range(0,len(s)):
            if s[i]=='(':
                stack.append(i)
            else:
                stack.pop()
               
                if len(stack)<1:
                    stack.append(i)
                else:
                    maxLen= max(maxLen,i-stack[-1])
        return maxLen
        